### ISSUE TITLE
CA1508: False Positive in 3.3.1. Older Version 3.3.0 is working.

### ISSUE THREAD

**Issue by SebastianSchumann** [dotnet/roslyn-analyzers#4383]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

**Version**: v3.3.1 (Latest)

### Describe the bug / Steps To Reproduce

The following code produces two warnings:

```
public static ReadOnlyDictionary<TKey, ReadOnlyCollection<TValue>> AsReadOnlyItems<TKey, TValue>(this IDictionary<TKey, IList<TValue>> self, IEqualityComparer<TKey> keyComparer = null)
{
    _ = self ?? throw new ArgumentNullException(nameof(self));

    keyComparer ??= (self as Dictionary<TKey, TValue>)?.Comparer;
    var copy = new Dictionary<TKey, ReadOnlyCollection<TValue>>(self.Count, keyComparer);
    foreach (var kvp in self)
    {
        if (kvp.Value is null)
        {
            copy.Add(kvp.Key, null);
            continue;
        }

        var readOnlyValue = kvp.Value as ReadOnlyCollection<TValue> ?? new ReadOnlyCollection<TValue>(kvp.Value);
        copy.Add(kvp.Key, readOnlyValue);
    }

    return new ReadOnlyDictionary<TKey, ReadOnlyCollection<TValue>>(copy);
}
```

There is one warning in line ``keyComparer ??= (self as Dictionary<TKey, TValue>)?.Comparer;`` that told me that ``self`` is never null.
The other warning is in line ``var readOnlyValue = kvp.Value as ReadOnlyCollection<TValue> ?? new ReadOnlyCollection<TValue>(kvp.Value);`` where ``kvp.Value`` is meant to be never null.

### Expected behavior

The warnings should not appear.

**The specified code is working in version 3.3.0 without any changes.**

### Actual behavior

A warning for CA1508 appears.

**Comment by RobSiklos** [comment]
Another repro:

```csharp
public class Issue4383_CA1508_DisposableIsNeverNull
{
    private class Dis : IDisposable
    {
        public void Dispose() { }
    }

    public void DoSomething()
    {
        for (int i = 0; i < 10; i++)
        {
            IDisposable disposable = new Dis();

            // CA1508: 'disposable' is never 'null'. Remove or refactor the condition(s) to avoid dead code.
            using (disposable)
            {
                // Do something
            }
        }
    }
}
```
